### ISSUE TITLE
crypto/mem_dbg.c does not build without OPENSSL_NO_CRYPTO_MDEBUG

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#9480]
Since about two weeks, when building with `enable-crypto-mdebug` I get 

```
gcc  -I. -Icrypto/include -Iinclude -Iproviders/common/include -Iproviders/common/ciphers -Icrypto  -DAES_ASM -DBSAES_ASM -DECP_NISTZ256_ASM -DGHASH_ASM -DKECCAK1600_ASM -DMD5_ASM -DOPENSSL_BN_ASM_GF2m -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_CPUID_OBJ -DPOLY1305_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DVPAES_ASM -DWHIRLPOOL_ASM -DX25519_ASM -fPIC -pthread -m64 -Wa,--noexecstack -rdynamic -Wall -O0 -g --std=gnu90 -DDEBUG_UNUSED -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wswitch -Wsign-compare -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Wmissing-prototypes -Wstrict-prototypes -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-3\"" -DMODULESDIR="\"/usr/local/lib/ossl-modules\""   -MMD -MF crypto/libcrypto-lib-mem_dbg.d.tmp -MT crypto/libcrypto-lib-mem_dbg.o -c -o crypto/libcrypto-lib-mem_dbg.o crypto/mem_dbg.c
crypto/mem_dbg.c:212:5: error: no previous prototype for 'CRYPTO_mem_debug_push' [-Werror=missing-prototypes]
 int CRYPTO_mem_debug_push(const char *info, const char *file, int line)
     ^
crypto/mem_dbg.c:217:5: error: no previous prototype for 'CRYPTO_mem_debug_pop' [-Werror=missing-prototypes]
 int CRYPTO_mem_debug_pop(void)
     ^
cc1: all warnings being treated as errors
Makefile:10240: recipe for target 'crypto/libcrypto-lib-mem_dbg.o' failed
make[1]: *** [crypto/libcrypto-lib-mem_dbg.o] Error 1
```

**Comment by levitte** [comment]
So basically, you can't enable crypto-mdebug

**Comment by DDvO** [comment]
Yes, or I remove `-Werror` form the Makefile or add a guard as follows to `mem_dbg.c`:

````
#ifdef OPENSSL_NO_CRYPTO_MDEBUG
int CRYPTO_mem_debug_push(const char *info, const char *file, int line)
{
    return 0;
}

int CRYPTO_mem_debug_pop(void)
{
    return 0;
}
#endif
```

which might be the solution, but looks counter-intuitive to me.

**Comment by richsalz** [comment]
Does wrapping the two functions inside ```#if !OPENSSL_API_3 ... #endif``` fix the problem?

**Comment by DDvO** [comment]
Yes, it does.

**Comment by mspncp** [comment]
@DDvO which revision are you compiling? I copy & pasted your command line and got no error on master (e9a5932d04f6b7dd25b39a8ff9dc162d64a78c22).

``` 
msp@msppc:~/src/openssl$ gcc  -I. -Icrypto/include -Iinclude -Iproviders/common/include -Iproviders/common/ciphers -Icrypto  -DAES_ASM -DBSAES_ASM -DECP_NISTZ256_ASM -DGHASH_ASM -DKECCAK1600_ASM -DMD5_ASM -DOPENSSL_BN_ASM_GF2m -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_CPUID_OBJ -DPOLY1305_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DVPAES_ASM -DWHIRLPOOL_ASM -DX25519_ASM -fPIC -pthread -m64 -Wa,--noexecstack -rdynamic -Wall -O0 -g --std=gnu90 -DDEBUG_UNUSED -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wswitch -Wsign-compare -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Wmissing-prototypes -Wstrict-prototypes -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-3\"" -DMODULESDIR="\"/usr/local/lib/ossl-modules\""   -MMD -MF crypto/libcrypto-lib-mem_dbg.d.tmp -MT crypto/libcrypto-lib-mem_dbg.o -c -o crypto/libcrypto-lib-mem_dbg.o crypto/mem_dbg.c
msp@msppc:~/src/openssl$ 
```

```
msp@msppc:~/src/openssl$ gcc --version
gcc (Gentoo 8.3.0-r1 p1.1) 8.3.0
```

**Comment by DDvO** [comment]
I'm using a very recent OpenSSL commit: 5d00f46e1527235ecd28a14c5413355c3ea66dfc
GCC version is pretty old (5.4.0) but I don't thinks this matters here.

**Comment by DDvO** [comment]
Also @Akretsch got the same error today, 
and for me the error occurs even after I've just switched to e9a5932.
@mspncp, strange that you cannot reproduce it.

**Comment by mspncp** [comment]
> and for me the error occurs even after I've just switched to e9a5932.
> @mspncp, strange that you cannot reproduce it.

Ok, now I understand (I hope): Did you configure using `--api=3.0.0` ? If I do it, I get the same error. Which means that @richsalz' suggestion is is probaby the right fix: It seems like function prototypes are guarded using a deprecation macro, but the function bodies are not.

**Comment by DDvO** [comment]
Yes, (essentially) this is key!
We did not explicitly configure with `--api=3.0.0`, but with `no-deprecated`,
and for this reason my `configdata.pm` contains `api => "3.0.0"`.
Then `make` inserts in `include/openssl/opensslconf.h` a `# define OPENSSL_MIN_API 3` (rather than `-1`),
and this explains the different behavior while the `gcc` command line is identical.

**Comment by mattcaswell** [comment]
This was fixed by #9483. Closing.
